File: Tree/P437PathSumIII/PathSumIII.py
import unittest

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def helper(root, map, presum, sum):
            nonlocal count
            if not root:
                return
            presum += root.val
            if presum - sum in map:
                count += map[presum - sum]
            if presum in map:
                map[presum] += 1
            else:
                map[presum] = 1
            helper(root.left, map, presum, sum)
            helper(root.right, map, presum, sum)
            map[presum] -= 1  
            
        if not root:
            return 0
        map = {}
        map[0] = 1
        count = 0
        helper(root, map, 0, sum)
        return count
        
 

class UnitTest(unittest.TestCase):
    def test_1(self):
        n1 = TreeNode(10)
        n2 = TreeNode(5)
        n3 = TreeNode(-3)
        n4 = TreeNode(3)
        n5 = TreeNode(2)
        n6 = TreeNode(11)
        n7 = TreeNode(3)
        n8 = TreeNode(-2)
        n9 = TreeNode(1)
        n1.left = n2
        n1.right = n3
        n2.left = n4
        n2.right = n5
        n3.right = n6
        n4.left = n7
        n4.right = n8
        n5.right = n9
        sum = 8
        self.assertEquals(3,  Solution().pathSum(n1, sum))

if __name__ == "__main__":
    unittest.main()     

File: Tree/P437PathSumIII/PathSumIII.java
package Tree.P437PathSumIII;

import java.util.HashMap;
import java.util.Map;

/**
 * @CLassName PathSumIII
 * @Description
 * You are given a binary tree in which each node contains an integer value.
 *
 * Find the number of paths that sum to a given value.
 *
 * The path does not need to start or end at the root or a leaf, but it must go downwards (traveling only from parent nodes to child nodes).
 *
 * The tree has no more than 1,000 nodes and the values are in the range -1,000,000 to 1,000,000.
 *
 * Example:
 *
 * root = [10,5,-3,3,2,null,11,3,-2,null,1], sum = 8
 *
 *       10
 *      /  \
 *     5   -3
 *    / \    \
 *   3   2   11
 *  / \   \
 * 3  -2   1
 *
 * Return 3. The paths that sum to 8 are:
 *
 * 1.  5 -> 3
 * 2.  5 -> 2 -> 1
 * 3. -3 -> 11
 * @Author cst
 * @Date 12/28/20  2:23 PM
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class PathSumIII {
    private int count = 0;
    public int pathSum(TreeNode root, int sum) {
        if (root == null) {
            return count;
        }
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        helper(root, 0, sum, map);
        return count;
    }
    public void helper(TreeNode root, int presum, int sum, Map<Integer, Integer> map) {
        System.out.println("Now map contains " + map);

        if (root == null) {
            return;
        }
        presum += root.val;
        if (map.containsKey(presum - sum)) {
            System.out.println("Map contains presum " + presum + " and sum " + sum);
            count += map.get(presum - sum);
        }
        map.put(presum, map.getOrDefault(presum, 0) + 1);
        helper(root.left, presum, sum, map);
        helper(root.right, presum, sum, map);
        map.put(presum, map.get(presum) - 1);
        presum -= root.val;
    }
}
